File: forms/diabetes-assessment/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/diabetes_grader.rs
use super::diabetes_rules::all_rules;
use super::types::{AssessmentData, ControlLevel, FiredRule};
use super::utils::{calculate_composite_score, collect_scored_items};

/// Pure function: evaluates all diabetes rules against assessment data.
/// Returns the control level, composite score (0-100), and fired rules.
/// If fewer than 3 scored items are answered, returns "draft" status.
pub fn calculate_control(data: &AssessmentData) -> (ControlLevel, f64, Vec<FiredRule>) {
    // Check if enough items are answered
    let items = collect_scored_items(data);
    let answered_count = items.iter().filter(|x| x.is_some()).count();

    // Also count HbA1c as a key data point
    let has_hba1c = data.glycaemic_control.hba1c_value.is_some();

    if answered_count < 2 && !has_hba1c {
        return ("draft".to_string(), 0.0, vec![]);
    }

    // Calculate composite score
    let score = calculate_composite_score(data).unwrap_or(0.0);

    // Fire rules
    let mut fired_rules = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                concern_level: rule.concern_level.to_string(),
            });
        }
    }

    // Determine control level from composite score
    let level = if score <= 20.0 {
        "veryPoor"
    } else if score <= 40.0 {
        "poor"
    } else if score < 65.0 {
        "suboptimal"
    } else {
        "wellControlled"
    };

    (level.to_string(), score, fired_rules)
}

File: forms/diabetes-assessment/front-end-form-with-svelte/src/lib/engine/diabetes-grader.ts
import type { AssessmentData, ControlLevel, FiredRule } from './types';
import { evaluateRules } from './diabetes-rules';
import { collectScoredItems, calculateCompositeScore } from './utils';

/**
 * Pure function: evaluates all diabetes rules against assessment data.
 * Returns the control level, composite score (0-100), and fired rules.
 * If fewer than 2 scored items are answered and no HbA1c, returns "draft" status.
 */
export function calculateControl(data: AssessmentData): {
	controlLevel: ControlLevel;
	controlScore: number;
	firedRules: FiredRule[];
} {
	// Check if enough items are answered
	const items = collectScoredItems(data);
	const answeredCount = items.filter((x) => x !== null).length;
	const hasHba1c = data.glycaemicControl.hba1cValue !== null;

	if (answeredCount < 2 && !hasHba1c) {
		return { controlLevel: 'draft', controlScore: 0, firedRules: [] };
	}

	// Calculate composite score
	const score = calculateCompositeScore(data) ?? 0;

	// Fire rules
	const firedRules = evaluateRules(data);

	// Determine control level from composite score
	let controlLevel: ControlLevel;
	if (score <= 20) {
		controlLevel = 'veryPoor';
	} else if (score <= 40) {
		controlLevel = 'poor';
	} else if (score < 65) {
		controlLevel = 'suboptimal';
	} else {
		controlLevel = 'wellControlled';
	}

	return { controlLevel, controlScore: score, firedRules };
}
